src/main/java/com/mongodb/hibernate/internal/type/MongoStructJdbcType.java

```diff
@@ -132,26 +153,60 @@ public BsonDocument createJdbcValue(Object domainValue, WrapperOptions options)
         return result;
     }
 
+    /**
+     * @return The {@linkplain Struct#getAttributes() struct attributes}. Though, the way we support {@link Struct} in
+     *     {@link MongoStructJdbcType} does not involve Hibernate ORM ever {@linkplain Connection#createStruct(String,
+     *     Object[]) creating} one. If we extended {@link org.hibernate.dialect.StructJdbcType}, this could have been
+     *     different.
+     */
     @Override
-    public Object[] extractJdbcValues(Object rawJdbcValue, WrapperOptions options) throws SQLException {
+    public Object @Nullable [] extractJdbcValues(@Nullable Object rawJdbcValue, WrapperOptions options)
+            throws SQLException {
+        if (isNull(rawJdbcValue)) {
+            throw new FeatureNotSupportedException(
+                    "TODO-HIBERNATE-48 https://jira.mongodb.org/browse/HIBERNATE-48 return null");
+        }
         if (!(rawJdbcValue instanceof BsonDocument bsonDocument)) {
             throw fail();
         }
+        var embeddableMappingType = getEmbeddableMappingType();
         var result = new Object[bsonDocument.size()];
         var elementIdx = 0;
         for (var value : bsonDocument.values()) {
-            assertNotNull(value);
-            result[elementIdx++] =
-                    value instanceof BsonDocument ? extractJdbcValues(value, options) : toDomainValue(value);
+            var jdbcMapping =
+                    embeddableMappingType.getJdbcValueSelectable(elementIdx).getJdbcMapping();
+            var jdbcTypeCode = jdbcMapping.getJdbcType().getJdbcTypeCode();
+            Object domainValue;
+            if (isNull(value)) {
+                throw new FeatureNotSupportedException(
+                        "TODO-HIBERNATE-48 https://jira.mongodb.org/browse/HIBERNATE-48 domainValue = null");
+            } else if (jdbcTypeCode == getJdbcTypeCode()) {
+                if (!(jdbcMapping.getJdbcValueExtractor() instanceof Extractor<?> structValueExtractor)) {
+                    throw fail();
+                }
+                domainValue = structValueExtractor.getJdbcType().extractJdbcValues(value, options);
+            } else if (jdbcTypeCode == MongoArrayJdbcType.JDBC_TYPE.getVendorTypeNumber()) {
+                if (!(jdbcMapping.getJdbcType() instanceof MongoArrayJdbcType arrayJdbcType)) {
+                    throw fail();
+                }
+                if (!(jdbcMapping.getJdbcValueExtractor() instanceof BasicExtractor<?> jdbcValueExtractor)) {
+                    throw fail();
```

Comment:
seems it is a good timing to introduce an assertion method to `MongoAssertions` to avoid the boilerplate:
```
    public static <T> T assertInstanceOf(Object value, Class<T> tClass) {
        if (!tClass.isInstance(value)) {
            throw new AssertionError();
        }
        return tClass.cast(value);
    }
```

Reply:
Done in ed6fb2c4b320eb5aa875a60772ba786ab64a3fcd.